### PR TITLE
Lint for unused borrows as part of `UNUSED_MUST_USE`

### LINKED ISSUE #76264: Warn when a borrow expression is unused
Borrowing an expression is side-effect free, so there should never be a reason to write the first instead of the second.

```rust
&expr;
// vs
expr;
```

If you really wanted to do this you could assign the result to an unused variable:

```rust
let _ = &expr;
```

Therefore, we should add a lint for unused borrows to the compiler. Why is this important? A misplaced semi-colon before an `&&` operator in a multi-line boolean expression causes subsequent clauses to be silently ignored.

```rust
let trex = TyrannosaurusRex::new();
let is_a_dog = has_a_tail(trex)
  && has_bad_breath(trex)
  && is_a_carnivore(trex); // Misplaced semi-colon (perhaps due to reordering of lines)
  && is_adorable(trex);

if is_a_dog {
    give_hug(trex); // Ouch!
}
```

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Resolves #76264.

This means an `unused_must_use` warning for statements like `&expr;` and `&mut expr;`. `unused_must_use` was chosen because it already lints for logical operators (`&&` and `||`) whose result is unused. Unused borrows actually appear fairly often in `rustc`'s test suite, since we have tests that rely on side-effects of the index operator (panicking). These cannot be written as `expr[..];`, since the result is unsized, but they can be written as `let _ = &expr[..];`, which is what this PR does.

See the linked issue for the motivating example. This lint also found a benign mistake in the derived impl of `HashStable`.

### CODE REVIEW

**Review by lcnr — APPROVED** [review]
LGTM

does this need lang or compiler signoff considering that this is a userfacing change?

**Comment by matthiaskrgr** [comment]
How similar is this to clippys needless_borrow lint?
https://rust-lang.github.io/rust-clippy/master/#needless_borrow

**Comment by ecstatic-morse** [comment]
I think it needs @rust-lang/lang approval yes.

@matthiaskrgr Probably the same. I don't use clippy.

**Comment by ecstatic-morse** [comment]
Ah, so the transformation from `&expr;` to `expr;` isn't always correct. If `expr` returns an unsized value like a slice, it will not compile.  I think we generally don't have lints in the compiler with false positives. However, I would really like to fix the pathological case in #76264 (reproduced below), a bare `&x[0..4];` should be pretty rare outside of the compiler's test suite and you can just use `let _ = &expr;`. I'll update the PR in any case.

```rust
let trex = TyrannosaurusRex::new();
let is_a_dog = has_a_tail(trex)
  && has_bad_breath(trex)
  && is_a_carnivore(trex); // Misplaced semi-colon (perhaps due to reordering of lines)
  && is_adorable(trex);

if is_a_dog {
    give_hug(trex); // Ouch!
}
```

**Comment by lcnr** [comment]
> If expr returns an unsized value like a slice, it will not compile (without unsized rvalues anyways).

Afaict the exprs which do so are the following:
```rust
&slice[..];
&y[0];
&*boxed;
```
all of which seem like something we can (and should) warn on, even if the suggestion is incorrect.

**Comment by ecstatic-morse** [comment]
For my own clarity, postfix indexing (`a[idx]`) desugars to `*std::ops::IndexMut::index_mut(&mut a, idx)` unless `a` is a slice or an array and `idx` is a `usize`, so conceptually the only problem is reborrows (`&*`). 

Also, I'm starting to think this should be folded into `UNUSED_MUST_USE`. The logical boolean operators are already included in that lint, and they have a similar "false-positive" problem, since someone could put a function with side-effects in the second half (e.g. `condition || panic!()`). The suggestion there is the same as for unused borrows: Assign the result to `_`.

**Comment by ecstatic-morse** [comment]
This is now part of `unused_must_use`, and I've updated the title and description. This is just waiting on @rust-lang/lang to weigh in.

**Comment by scottmcm** [comment]
This makes sense to me.  If we lint on `!x;`, we might as well lint on `&x` too.

I don't know whether this needs a full FCP (we can talk about it on Monday), but in case it does,
@rfcbot fcp merge

**Comment by ecstatic-morse** [comment]
> I don't know whether this needs a full FCP (we can talk about it on Monday), but in case it does,

I definitely approved a similar PR (#74869), without lang team approval. https://github.com/rust-lang/rust/pull/76894#pullrequestreview-491774016 asked, and it occurred to me that I didn't actually know. If the lang-team needs to fill some time, maybe they want to weigh in on #69466, which also involves side-effects and `UNUSED_MUST_USE`. I'm thinking their might be better things to do, however.

**Comment by withoutboats** [comment]
@rfcbot concern crater

Enthusiastic about this intuitively, but I'd like a crater run to make sure that this isn't going to suddenly lint a large part of the ecosystem or anything.

**Comment by joshtriplett** [comment]
Seems very reasonable. Lang normally needs to :+1: new lints. We discussed this in the @rust-lang/lang meeting today, and the consensus was that we're in favor of this as proposed, as long as there isn't some common pattern we're unaware of that this will flag.

**Comment by scottmcm** [comment]
@craterbot check

**Comment by ecstatic-morse** [comment]
@scottmcm, do you have a plan for triaging the crater results? I've never done one for warnings, but I suspect the usual approach is to run with `-D the_lint`. However, this currently folded into `unused_must_use`, and I suspect denying this will significantly reduce the corpus.

**Comment by taiki-e** [comment]
Seems many of the regressions is due to that this lint warning the code generated by external macros.

**Comment by ecstatic-morse** [comment]
About 15000 new warnings in the ecosystem.

```sh
> rg 'unused borrow that must be used' | wc -l
14598
```

Many are in generated code:

```sh
> rg -A6 'unused borrow that must be used' | rg 'pin_project' | wc -l
4226
```
About a third are due to `pin-project`

```sh
rg -A6 'unused borrow that must be used' | rg '[A-Z]+::ptr\(\)' | wc -l
1941
```
At least an eighth from embedded-hal. For example:
```rust
&(*$GPIOX::ptr()).moder.modify(|r, w|
    w.bits((r.bits() & !(0b11 << offset)) | (input << offset))
);
```

```shell
> rg -A6 'unused borrow that must be used' | rg 'PackedStruct' | wc -l
594
```
A few more are due to `PackedStruct`.

The rest are legitimate regressions, usually due to a misunderstanding of the
autoref rules. Most are benign but unnecessary (e.g. `&mut foo.rotate()`), and I think users would benefit from learning that their `&` annotations aren't doing what they think. This is especially true in the following case :smile:.

```rust
&vec1.push(0);
```

```shell
> rg -A6 'unused borrow that must be used' | rg '&(mut)?\s*\w+\[' | wc -l
1020
```
Putting `&` before calling a method on a slice seems to be pretty common. Presumably a bare slice looks strange to some users
```rust
slice[..].method()
```

```sh
> rg -A6 'unused borrow that must be used' | rg --case-sensitive '\.\.\w*\];' | wc -l
60
```
There's also a few cases where the `&foo[..];` pattern appears, and these will need to use `let _ = ...` instead of simply removing the borrow.

**Comment by crlf0710** [comment]
Triage: what's the next steps here?

**Comment by lcnr** [comment]
hmm

> We should probably add a help message to this effect.

Seems like a good idea

> There's also a few cases where the &foo[..]; pattern appears, and these will need to use let _ = ... instead of simply removing the borrow.

Can we add a help for this to the lint?

We may also want to not emit it inside of macros for now, though I am not too sure how easy it is to start linting there again after the lint was in stable for a while :thinking:

@ecstatic-morse what's you opinion here? Once you are satisfied we should probably once again nominate for `t-lang` to get a final signoff

**Comment by ecstatic-morse** [comment]
@withoutboats a crater run has been completed and triaged. Where do we go from here?

**Comment by lcnr** [comment]
nominating for `T-lang` again. IMO the fallout of this PR is small enough for us to go ahead here. Especially considering that it helps with catching actual user errors.

**Comment by nikomatsakis** [comment]
We discussed this in our [2021-01-26](https://github.com/rust-lang/lang-team/blob/master/minutes/2021-01-26.md) meeting. Apologies for being slow to write that up! 

We had a few thoughts and questions:

* First, @ecstatic-morse, do we know why pin-project is getting warnings? That seems worrisome.
* What about embedded-hal?

My question:

* Is there still a false warning in the case of a value of unsized type?

**Comment by withoutboats** [comment]
@rfcbot resolve crater

**Comment by nikomatsakis** [comment]
@rfcbot fcp cancel

It seems like there's a bug in the RFC bot. I'm going to cancel the FCP and restart it to get things going on.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

**Comment by nikomatsakis** [comment]
@rfcbot concern crater-run-warnings

I'm repeating my questions to @ecstatic-morse or others. Do we know why pin-project is getting warnings? What about embedded-hal?

**Comment by ecstatic-morse** [comment]
`pin-project` generates a function which borrows each field of a struct to assert that it is not `#[repr(packed)]`. Newer versions of `pin-project` [now emit `let _ = ...`](https://github.com/taiki-e/pin-project/commit/759432193d3cac83cfc1cf3ee33dc4e6a81293b5#diff-e6e83338c6ca03fbdb451810d3b9945db7acf3de8fcff6ca94676882703300d9L106) and will no longer cause warnings with this change.

As I said in the summary, `embedded-hal` generates code like:

```rust
&(*$GPIOX::ptr()).moder.modify(|r, w|
    w.bits((r.bits() & !(0b11 << offset)) | (input << offset))
);
```

This code works due to the autoref rules, but it seems the author thought the `&` was necessary and that it would associate with the field access (`moder`) but not with the method call (`modify`). This is not what happens, the `&` is applied to the result of `modify` and is therefore useless, so I think the warning is helpful in this case. These warnings will annoy users of code generated by `embedded-hal` until it is updated, but I don't think this should block these changes.

**Comment by nikomatsakis** [comment]
@ecstatic-morse thanks for the writeup! I'm satisfied by these comments. I'm going to resolve the concern.

@rfcbot resolve crater-run-warnings

File: compiler/rustc_macros/src/hash_stable.rs [hunk 1]
```diff
@@ -54,7 +54,7 @@ pub fn hash_stable_generic_derive(mut s: synstructure::Structure<'_>) -> proc_ma
             quote! {}
         } else if let Some(project) = attrs.project {
             quote! {
-                &#bi.#project.hash_stable(__hcx, __hasher);
+                (&#bi.#project).hash_stable(__hcx, __hasher);
```
Comment (cjgillot) [inline compiler/rustc_macros/src/hash_stable.rs:57]: Given how long this mistake has been there without consequence, just removing the borrow should be fine.

**Comment by Dylan-DPC** [comment]
@bors r- 

failed in [rollup](https://github.com/rust-lang/rust/pull/82600)

**Comment by Dylan-DPC** [comment]
@ecstatic-morse any updates on this?

**Comment by lcnr** [comment]
I am not able to review any PRs in the near future.

r? @RalfJung

**Comment by RalfJung** [comment]
Given that @ecstatic-morse has been inactive here for months and in general does not seem to be doing much Rust work any more, I'm going to close this PR for now.
@ecstatic-morse if you (or anyone else) wants to pick this up again, feel free to reopen. :)

**Pull request closed without merging**